### PR TITLE
Added class representing a Salt Formula Form

### PR DESCRIPTION
**Pull request by teclator**

### CODE REVIEW

File: src/lib/y2configuration_management/salt/form.rb [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+      end
+
+      # It builds the form elements from the specification
+      def build_elements
+        spec_elements.each { |n, h| elements << FactoryFormElement.build(n, h, parent: self) }
+      end
+    end
+
+    # It builds new Form Elements depending on its specification type
+    class FactoryFormElement
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: This should be named `FormElementFactory`, or even `FormElement::Factory`.

File: src/lib/y2configuration_management/salt/form.rb [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yaml"
+
+module Y2ConfigurationManagement
+  module Salt
+    # Class that represents a form for Salt Formulas
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: IMHO we should link to explanations, to help non-experts who stumble across the class:

> 
```suggestion
    # A [Form][1] for [Salt Formulas][2].
    # [1]: https://www.suse.com/documentation/suse-manager-3/3.2/susemanager-best-practices/html/book.suma.best.practices/best.practice.salt.formulas.and.forms.html#best.practice.salt.formulas.pillar
    # [2]: https://docs.saltstack.com/en/latest/topics/development/conventions/formulas.html
```

File: src/lib/y2configuration_management/salt/form.rb [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    class Form
+      # @return [Array<FormElement>]
+      attr_reader :elements
+      # @return [Hash]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: Is this right? 

```suggestion
      # The original specification (deserialized form.yml).
      # @return [Hash]
```

File: src/lib/y2configuration_management/salt/form.rb [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+      end
+    end
+
+    # Three different kind of elements:
+    #
+    # scalar values, groups and collections
+    class FormElement
+      # @return [Form, FormElement]
+      attr_reader :parent
+      # @return  [String]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: Is this the internal name (pillar key) or the user visible name (`$name`)?

File: src/lib/y2configuration_management/salt/form.rb [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+      attr_reader :name
+      # @return [String]
+      attr_reader :help
+      # @return scope [String] specify the level in which the value can be
+      #   edited. Possible values are: system, group and readonly
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: `$scope` and `$type` are enumerations. Wouldn't it be better to represent them as Symbols internally?

File: src/lib/y2configuration_management/salt/form.rb [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+          FormInput
+        end
+      end
+    end
+
+    # Three different kind of elements:
+    #
+    # scalar values, groups and collections
+    class FormElement
+      # @return [Form, FormElement]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: Returning one of two unrelated classes is a code smell. I think that `Form` is in fact a subclass of `Container`.

Alternatively, the `Form` could contain a `root` attribute which would be a `Container`.

File: src/lib/y2configuration_management/salt/form.rb [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+      # Constructor
+      #
+      # @param name [String]
+      # @param spec [Hash] form element specification
+      def initialize(name, spec, parent: nil)
+        @name = name
+        @type = spec["$type"] || "text"
+        @help = spec["$help"] if spec ["$help"]
+        @scope = spec["$scope"] if spec["$scope"]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: I think we should spell out the defaults explicitly,
```suggestion
        @scope = spec.fetch("$scope", "system")
```

File: src/lib/y2configuration_management/salt/form.rb [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+      #   edited. Possible values are: system, group and readonly
+      attr_reader :scope
+      # @return optional [Boolean]
+      attr_reader :optional
+
+      # Constructor
+      #
+      # @param name [String]
+      # @param spec [Hash] form element specification
+      def initialize(name, spec, parent: nil)
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: IMHO leaving out the parent is an exception, not the usual case, so it should not have a default (but rather the root element should explicitly say `parent:nil`.
```suggestion
      def initialize(name, spec, parent:)
```

File: src/lib/y2configuration_management/salt/form.rb (hunk 6 above)
Comment (teclator) [reply src/lib/y2configuration_management/salt/form.rb]: Yep, completely agree, I mostly added for being able to see also your siblings thinking in the `visibleIf` option but probably that is a responsibility for the UI and the form but not for the Form elements.

File: src/lib/y2configuration_management/salt/form.rb (hunk 3 above)
Comment (teclator) [reply src/lib/y2configuration_management/salt/form.rb]: I think so, Isn't it a hash? 

https://github.com/yast/yast-configuration-management/blob/master/test/fixtures/formulas/test-formula/form.yml#L1

File: src/lib/y2configuration_management/salt/form.rb (hunk 4 above)
Comment (teclator) [reply src/lib/y2configuration_management/salt/form.rb]: Currently it is always the `pillar key` although it should be the `$name` or the `pillar key. So, we probably should have both, one as the ID and the other as the visible name.

@imobachgs what do you think?

File: test/lib/y2configuration_manager/salt/form_test.rb [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+  end
+
+  describe "#path" do
+    let(:formula_path) { FORMULAS_PATH.join("test-formula") }
+    let(:form_path) { formula_path.join("form.yml") }
+    let(:form) { Y2ConfigurationManagement::Salt::Form.from_file(form_path) }
+
+    it "returns the absolute form element path in the Form" do
+      number = form.find_element_by(name: "number")
+      expect(number.path).to eql(".root.demo.number")
```
Comment (mvidner) [inline test/lib/y2configuration_manager/salt/form_test.rb]: Is this "path" something defined elsewhere, or a new thing invented now?

If it's new, we should use something isomorphic but standardized: [JSON Pointer](https://tools.ietf.org/html/rfc6901)

File: src/lib/y2configuration_management/salt/form.rb [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+        end
+      end
+    end
+
+    # Three different kind of elements:
+    #
+    # scalar values, groups and collections
+    class FormElement
+      PATH_DELIMITER = ".".freeze
+      # @return [String]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]:
```suggestion
      # @return [String] The key for the pillar
```

File: src/lib/y2configuration_management/salt/form.rb [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+      attr_reader :id
+      # @return [Symbol]
+      attr_reader :type
+      # @return [FormElement]
+      attr_reader :parent
+      # @return  [String]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]:
```suggestion
      # @return  [String] The user visible name ($name)
```

File: src/lib/y2configuration_management/salt/form.rb [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+      attr_reader :help
+      # @return [Symbol] specify the level in which the value can be edited.
+      #   Possible values are: system, group and readonly
+      attr_reader :scope
+      # @return optional [Boolean]
+      attr_reader :optional
+
+      # Constructor
+      #
+      # @param name [String]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: Rename `name` to `id` in the whole constructor

File: src/lib/y2configuration_management/salt/form.rb [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+        @optional = spec["$optional"] if spec["$optional"]
+        @parent = parent
+      end
+
+      # Return the absolute path of this form element in the actual form
+      #
+      # FIXME: possible implementation of the form element path
+      #
+      # @return [String]
+      def path
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: Is this the `$visibleIf` path like `..otherGroup#otherWidget` ?

File: src/lib/y2configuration_management/salt/form.rb [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+      # @return [String] help text usually displayed in the input field
+      attr_reader :placeholder
+      # @return [Boolean, Integer, String, nil] default input value
+      attr_reader :default
+      # @return [Array<String>] a list of possible values for a select input
+      attr_reader :values
+
+      # Constructor
+      #
+      # @param name [String]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: id

File: src/lib/y2configuration_management/salt/form.rb [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+      end
+    end
+
+    # Scalar value FormElement
+    class FormInput < FormElement
+      # @return [String] help text usually displayed in the input field
+      attr_reader :placeholder
+      # @return [Boolean, Integer, String, nil] default input value
+      attr_reader :default
+      # @return [Array<String>] a list of possible values for a select input
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: the types for these 3 attrs should be unified

File: src/lib/y2configuration_management/salt/form.rb [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+    end
+
+    # Container Element
+    class Container < FormElement
+      # @return elements [Array<FormElement>]
+      attr_reader :elements
+
+      # Constructor
+      #
+      # @param name [String]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: id

File: src/lib/y2configuration_management/salt/form.rb [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+      #
+      #   f = Y2ConfigurationManagemenet.from_file("form.yml")
+      #   f.find_element_by(name: "subnets") #=> <Collection @name="subnets"
+      #   f.find_element_by(path: ".root.dhcpd") #=> <Container @name="dhcpd"
+      #
+      # @param arg [Hash]
+      # @return [FormElement, nil]
+      def find_element_by(arg)
+        elements.each do |element|
+          return element if arg.any? { |k, v| element.public_send(k) == v }
```
Comment (imobachgs) [inline src/lib/y2configuration_management/salt/form.rb]: I would prefer that a given instance check whether it meets the condition. Having that into account, I would write something like this (maybe I am overlooking something):

```ruby
return self if arg.any? { |k,v| public_send(k) == v } # is it me?
return nil if elements.empty? # you have come to the end
elements.find { |e| e.find_element_by(arg) }
```

File: src/lib/y2configuration_management/salt/form.rb [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+        @item_name = spec["item_name"] if spec["item_name"]
+        @min_items = spec["$minItems"] if spec["$minItems"]
+        @max_items = spec["$maxItems"] if spec["$maxItems"]
+        @prototype = prototype_for(name, spec)
+        @default = spec["$default"] if spec["$default"]
+      end
+
+    private
+
+      # Return a single or group of FormElements based on the prototype given
```
Comment (imobachgs) [inline src/lib/y2configuration_management/salt/form.rb]: I would write "`FormElement` objects" (not pluralizing class names). Or just "form elements".

File: src/lib/y2configuration_management/salt/form.rb [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+
+      # @param spec [Hash] form element specification
+      def build_elements(spec)
+        spec.select { |k, v| !k.start_with?("$") }.each do |name, spec|
+          @elements << FormElementFactory.build(name, spec, parent: self)
+        end
+      end
+    end
+
+    # Defines a collection of FormElements or Containers all of them based in
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]:
```suggestion
    # Defines a collection of {FormElement}s or {Container}s all of them based in
```

File: src/lib/y2configuration_management/salt/form.rb [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+        new(definition)
+      end
+
+      # Convenience method for looking for a particular FormElement.
+      def find_element_by(arg)
+        root.find_element_by(arg)
+      end
+    end
+
+    # It builds new Form Elements depending on its specification type
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]:
```suggestion
    # It builds new {FormElement}s depending on its specification type
```

File: src/lib/y2configuration_management/salt/form.rb [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+    # the same prototype.
+    class Collection < FormElement
+      # @return [Integer] lowest number of elements that needs to be defined
+      attr_reader :min_items
+      # @return [Integer] highest number of elements that needs to be defined
+      attr_reader :max_items
+      # @return [String] name for the members of the collection
+      attr_reader :item_name
+      # list of elements (let's see if we promote it to a class)
+      # or children or whatever:xs:xa
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: xs xa ??

File: src/lib/y2configuration_management/salt/form.rb [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+      attr_reader :prototype
+
+      # Default collection values
+      attr_reader :default
+
+      # Constructor
+      #
+      # @param name [String]
```
Comment (mvidner) [inline src/lib/y2configuration_management/salt/form.rb]: id

File: src/lib/y2configuration_management/salt/form.rb (hunk 21 above)
Comment (mvidner) [reply src/lib/y2configuration_management/salt/form.rb]: Anyway, I haven't checked the details, but wouldn't it come out easier if we said that the prototype was always  a Container?

**Comment by mvidner** [comment]
Please fix RuboCop issues before merging, AFAICT they are easy.

File: src/lib/y2configuration_management/salt/form.rb (hunk 17 above)
Comment (imobachgs) [reply src/lib/y2configuration_management/salt/form.rb]: Ok, find is not the right iterator, but you get the idea :)

File: src/lib/y2configuration_management/salt/form.rb (hunk 12 above)
Comment (teclator) [reply src/lib/y2configuration_management/salt/form.rb]: Yep, already realized it, thnx

File: src/lib/y2configuration_management/salt/form.rb (hunk 13 above)
Comment (teclator) [reply src/lib/y2configuration_management/salt/form.rb]: It was mainly added taking that option into account

File: src/lib/y2configuration_management/salt/form.rb (hunk 21 above)
Comment (teclator) [reply src/lib/y2configuration_management/salt/form.rb]: We were already discussing this, and probably will be better to always have a container as we did in the Form class even for simple text fields.

**Pull request merged**